### PR TITLE
MOTECH-1956: sending xml template unit tests

### PR DESCRIPTION
**Pull request by matkwiatkowski**

### CODE REVIEW

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Test(expected = UnknownHostException.class)
+    public void shouldCreateCorrectPostMethod() throws Exception {
+
+        whenNew(PostMethod.class).withNoArguments().thenReturn(postMethod);
+        whenNew(HttpClient.class).withNoArguments().thenReturn(httpClient);
+        when(httpClient.executeMethod(postMethod)).thenReturn(200);
+
+        iheTemplateService.sendTemplateToRecipientUrl(new String("http://someURL:80/xml"), "someTemplate");
```
Comment (martokarski) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImplTest.java]: Again explicity constructor call. These are the same field as above?

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    public void setUp() {
+        initMocks(this);
+    }
+
+    @Test(expected = UnknownHostException.class)
+    public void shouldFailToConnect() throws IOException {
+        iheTemplateService.sendTemplateToRecipientUrl(url, template);
+    }
+
+    @Test(expected = UnknownHostException.class)
```
Comment (martokarski) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImplTest.java]: This should create a correct request, so why an exception is expected?

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.when;
+import static org.mockito.MockitoAnnotations.initMocks;
+import static org.powermock.api.mockito.PowerMockito.mock;
+import static org.powermock.api.mockito.PowerMockito.whenNew;
+
+@RunWith(PowerMockRunner.class)
+public class IHETemplateServiceImplTest {
+
+    private final String url = new String("http://someURL:80/xml");
```
Comment (martokarski) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImplTest.java]: Why do you explicity call String constructor?

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java [hunk 4]
```diff
@@ -61,4 +66,25 @@ public void shouldThrowExceptionWhenRecipientNotFound() throws ClassNotFoundExce
...
+            byteObjects[i++] = b;
+        }
+
+
+        when(iheTemplateDataService.findByName("sampleTemplate3")).thenReturn(cdaTemplate);
+        when(hl7RecipientsService.getRecipientbyName("sampleRecipient3")).thenReturn(hl7Recipient);
+        when((Byte[]) iheTemplateDataService.getDetachedField(cdaTemplate, "templateData")).thenReturn(byteObjects);
+
+        iheActionHelper.handleAction(params);
+    }
```
Comment (martokarski) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java:93]: Maybe use verify() clause to verify if method was called with proper parameters.

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java [hunk 5]
```diff
@@ -80,10 +83,10 @@ public void shouldCallSendTemplateMethod() throws IOException, SAXException, Par
             byteObjects[i++] = b;
         }
 
-
         when(iheTemplateDataService.findByName("sampleTemplate3")).thenReturn(cdaTemplate);
         when(hl7RecipientsService.getRecipientbyName("sampleRecipient3")).thenReturn(hl7Recipient);
         when((Byte[]) iheTemplateDataService.getDetachedField(cdaTemplate, "templateData")).thenReturn(byteObjects);
+        verify(iheTemplateService).sendTemplateToRecipientUrl(new String(bytes, StandardCharsets.UTF_8), new PostMethod(hl7Recipient.getRecipientUrl()));
```
Comment (martokarski) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java]: Why is this verify() before method call?  It should throw an exception here. I checked this and there is some problem with hl7Recipient.getRecipientUrl() - without this clause the "Wanted but not invoked" exception appears.

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java [hunk 6]
```diff
@@ -61,4 +69,25 @@ public void shouldThrowExceptionWhenRecipientNotFound() throws ClassNotFoundExce
...
+        Byte[] byteObjects = new Byte[bytes.length];
+        int i = 0;
+        for (byte b : bytes) {
+            byteObjects[i++] = b;
+        }
+
+        when(iheTemplateDataService.findByName("sampleTemplate3")).thenReturn(cdaTemplate);
+        when(hl7RecipientsService.getRecipientbyName("sampleRecipient3")).thenReturn(hl7Recipient);
+        when((Byte[]) iheTemplateDataService.getDetachedField(cdaTemplate, "templateData")).thenReturn(byteObjects);
+        verify(iheTemplateService).sendTemplateToRecipientUrl(new String(bytes, StandardCharsets.UTF_8), new PostMethod(hl7Recipient.getRecipientUrl()));
```
Comment (martokarski) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java]: This should be called after iheActionHelper.handleAction(params). This should return exception here. I checked this and the problem is with hl7Recipient.getRecipientUrl() - without this, the test throw an exception "Wanted but not invoked".

**Comment by martokarski** [comment]
Avoid changing code which was checked and merged already.

**Comment by martokarski** [comment]
@kmadej Looks good for me

**Comment by kmadej** [comment]
test this please

**Comment by kmadej** [comment]
test this please

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java [hunk 7]
```diff
@@ -61,4 +68,26 @@ public void shouldThrowExceptionWhenRecipientNotFound() throws ClassNotFoundExce
...
         iheActionHelper.handleAction(params);
     }
+
+    @Test
+    public void shouldCallSendTemplateMethod() throws IOException, SAXException, ParserConfigurationException, ClassNotFoundException {
+        Map<String, Object> params = new HashMap<>();
+        params.put(Constants.TEMPLATE_NAME_PARAM, "sampleTemplate3");
+        params.put(Constants.RECIPIENT_NAME_PARAM, "sampleRecipient3");
+
+        byte[] bytes = IOUtils.toByteArray(getClass().getResourceAsStream("/empty_template.xml"));
```
Comment (pgesek) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java]: always close streams

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java [hunk 8]
```diff
@@ -61,4 +68,26 @@ public void shouldThrowExceptionWhenRecipientNotFound() throws ClassNotFoundExce
...
+        Byte[] byteObjects = new Byte[bytes.length];
```
Comment (pgesek) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java]: Use ArrayUtils.toObject()

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java [hunk 9]
```diff
@@ -75,12 +77,10 @@ public void shouldCallSendTemplateMethod() throws IOException, SAXException, Par
...
 
-        byte[] bytes = IOUtils.toByteArray(getClass().getResourceAsStream("/empty_template.xml"));
-        Byte[] byteObjects = new Byte[bytes.length];
-        int i = 0;
-        for (byte b : bytes) {
-            byteObjects[i++] = b;
-        }
+        InputStream inputStream = getClass().getResourceAsStream("/empty_template.xml");
+        byte[] bytes = IOUtils.toByteArray(inputStream);
+        inputStream.close();
```
Comment (pgesek) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelperTest.java]: streams should be closed in finally. In java 7 you can do: 

try (InputStream is = ...) {

}

to achieve the same

**Pull request merged**